modes.go

```diff
@@ -109,6 +109,9 @@ func (ti *TestIterator) IsDone() bool {
 	}
 }
 
+var totalErrors int = 0
+var totalErrorsMsgPrinted bool = false
```

Comment:
The `RunTest` function reads and modifies those variables without any synchronization, which is wrong because `RunTest` is being run concurrently from multiple goroutines that might get scheduled to run from multiple OS threads. You need to properly synchronize these operations or you may get weird bugs in form of errors being improperly counted or the error message being printed more than once.

I would recommend using atomics here to operate on `totalErrors`. Look at the [atomic](https://pkg.go.dev/sync/atomic) package in the Go's standard library. For example, to increment `totalErrors` you should use `AddInt32(&totalErrors, 1)`. To read the variable, you should use e.g. `LoadInt32(&totalErrors)`.

As for `totalErrorsMsgPrinted`, you could just use [sync.Once](https://pkg.go.dev/sync#Once) to make sure that you print the error message at most once. It has a simpler interface than atomics and is easier to use correctly.

Reply:
Thanks for the review.

The `sync.Once` solves that small nit where we may print more than once the critical error, will reuse it.

But the `atomic.AddInt32` doesn't help, because the concurrency lays in different layer than adding value.
It gets updated lots of times before reaching the condition by any of the threads.
But it is not really a problem, it doesn't harm us anyhow if the counter gets bigger than the limit.
The most valuable thing that is already done - making decision right after the counter reaches the limit.

Reply:
> But the `atomic.AddInt32` doesn't help, because the concurrency lays in different layer than adding value. It gets updated lots of times before reaching the condition by any of the threads. But it is not really a problem, it doesn't harm us anyhow if the counter gets bigger than the limit. The most valuable thing that is already done - making decision right after the counter reaches the limit.

I agree that it's not a problem if `totalErrors` gets bigger than the limit and we don't react immediately. That's not the problem the atomics are supposed to solve here. The main problem is that `totalErrors += 1` is not an atomic, indivisible operation. It usually consists of more than one step:

1. load the value of totalErrors from memory to a CPU register
2. increment the value in the register
3. store the value of the register to totalErrors

A goroutine may be paused between any of those instructions and another may be executed in the meantime. For example, if you have two goroutines trying to increment `totalErrors` like this, you may get the following race:

```
totalErrors is initially zero

A: registerA <- totalErrors

# Goroutine A pauses, goroutine B is executed now

B: registerB <- totalErrors
B: registerB += 1
B: totalErrors <- registerB

# Goroutine B pauses, goroutine A is executed now

A: registerA += 1
A: totalErrors <- registerA

totalErrors is now set to 1
```

Although you were supposed to increment the value twice, its value only gets larger by one. Unless you synchronize those operations somehow (e.g. via mutex or atomics) you may actually undercount.

In general: if you have a data race then weird things can happen, so you should avoid data races and synchronize everything properly.

Reply:
Done